src/auslib/web/public/updatetrack.py

```diff
@@ -0,0 +1,36 @@
+from flask import g, request
+
+
+class UpdateTrack:
+    def __init__(self, url):
+        self.url = url
+        # split url components...
+
+        self.release_name = None
+        self.is_fallback_mapping = False
+        self.update_xml = None
+        # others intermediate values
+
+    @staticmethod
+    def set_release_name(release_name):
+        g.update_track.release_name = release_name
+
+    @staticmethod
+    def set_is_fallback_mapping(is_fallback_mapping):
+        g.update_track.is_fallback_mapping = is_fallback_mapping
+
+    @staticmethod
+    def set_update_xml(xml):
+        g.update_track.update_xml = xml
+
+
+def configure_app(app):
+    @app.before_request
+    def before_request():
+        g.update_track = UpdateTrack(request.full_path)
```

Comment:
Hmm, I'm not sure I understand the need for the `UpdateTrack` object. As far as I know, everything we need to send to BigQuery is available in `after_request`, so we should just be able to do everything there...

...however, `after_request` blocks the request from returning, which means our update responses will be slower, and will potentially break if the submission to BigQuery does not work.

Can you look into whether it's possible to do the BigQuery submission after the request returns?

Reply:
If you are interested only in url parameters and the result XML, it is OK to collect this data on `after_request`.

But if do you want to include intermediate values, evaluated on request process, UpdateQuery object can keep this data until `after_request` reached.

Example here:

https://github.com/mozilla-releng/balrog/blob/86c29f36e2045f085c472eefe0d59306d04027fa/tests/web/test_updatetrack.py#L19-L24

The lines 21 and 22, captures values evaluated in the view function (between `before_request` and `after_request`).


> Can you look into whether it's possible to do the BigQuery submission after the request returns?

Yeah, I realized this problem too, I'm looking for if it is possible trigger an event outside Request context.


Let me know if I was clear about `UpdateTrack`, if the intermediate values are not necessary, no point to keep the object.

Reply:
> Let me know if I was clear about UpdateTrack, if the intermediate values are not necessary, no point to keep the object.

I think you're right about the intermediate values; mapping & fallback mapping are not available after the request has been handled, so we will need to store them partway through handling the request. Do you have any sense which parts of the existing code will end up calling `UpdateTrack`s setters?

> Yeah, I realized this problem too, I'm looking for if it is possible trigger an event outside Request context.

Let me know if you hit an impass here and we can look at it together.

Reply:
> Do you have any sense which parts of the existing code will end up calling UpdateTracks setters?

Probably in some place at `client.get_update_blob` or `AUS.evaluateRules` when the flow is in a request context.

> Let me know if you hit an impass here and we can look at it together.

I'm looking for a pure flask way, before look into other solutions. My next step will be test flask signals like `request_tearing_down` and `appcontext_tearing_down`.

Reply:
> Probably in some place at client.get_update_blob or AUS.evaluateRules when the flow is in a request context.

Sounds good. Let's keep it in `get_update_blob` if we can, since that's the view code.